File: zig/plugins/all.zig
//! THE FULL SET - every plugin this library ships, in one module.
//!
//! It exists for the callers that genuinely want all ten kinds: the CLI,
//! the conformance suite, an app whose chain is decided at run time.
//!
//!     const plugins = @import("sekretoplugins");
//!     const secrets = try sekreto.Sekreto.init(alloc, config, .{
//!         .plugins = &plugins.ALL,
//!         .providers = chain,
//!     });
//!
//! IT IS ALSO THE THING TO AVOID IF YOU CARE ABOUT SIZE. A module rooted
//! here compiles every plugin - AWS request signing and seven HTTPS vault
//! clients included - which is the cost the core/plugin split exists to
//! remove. A lean consumer roots its plugins module at the one file it
//! needs instead: `-Msekretoplugins=plugins/hashicorp.zig` compiles
//! hashicorp and `httpjson.zig` and nothing else, because zig analyses
//! only what a root reaches. See docs/design/plugin-providers.md.
//!
//! Also here, for whoever writes a plugin or a CLI: the shared HTTP
//! helper and the SigV4 signer.

const sekreto = @import("sekreto");

pub const httpjson = @import("httpjson.zig");
pub const sigv4 = @import("sigv4.zig");

pub const hashicorp = @import("hashicorp.zig").hashicorp;
pub const boru = @import("boru.zig").boru;
pub const awssecrets = @import("aws.zig").awssecrets;
pub const awsparams = @import("aws.zig").awsparams;
pub const gcpsecrets = @import("gcpsecrets.zig").gcpsecrets;
pub const azuresecrets = @import("azuresecrets.zig").azuresecrets;
pub const onepassword = @import("onepassword.zig").onepassword;
pub const doppler = @import("doppler.zig").doppler;
pub const infisical = @import("infisical.zig").infisical;
pub const secretspec = @import("secretspec.zig").secretspec;

/// Every plugin definition this library ships.
pub const ALL = [_]sekreto.Definition{
    hashicorp,    boru,        awssecrets, awsparams, gcpsecrets,
    azuresecrets, onepassword, doppler,    infisical, secretspec,
};

File: swift/plugins/All.swift
// THE FULL SET - every plugin this library ships, in one name.
//
// It exists for the callers that genuinely want all ten kinds: the CLI,
// the conformance suite, an app whose chain is decided at run time.
//
//     import SekretoPlugins
//     let secrets = try makesekreto(chain, plugins: allplugins)
//
// IT IS ALSO THE THING TO AVOID IF YOU CARE ABOUT SIZE - though in swift
// the cost is paid a module at a time rather than a name at a time.
// Linking `libSekretoPlugins.a` links AWS request signing and seven HTTP
// vault clients whether or not `allplugins` is ever mentioned, because a
// swift module is compiled and archived whole. A lean consumer therefore
// builds a SMALLER MODULE from the files it wants rather than reaching for
// a smaller name from this one:
//
//     swiftc -emit-module -emit-library -static -module-name SekretoPlugins \
//       plugins/Hashicorp.swift plugins/Httpjson.swift
//
// `make lean` does exactly that, and it is a test as well as an example:
// each plugin file must compile beside Httpjson.swift and nothing else, so
// a plugin that quietly reached into its neighbour - which one swift
// module would let it do with no import to give it away - fails there.
//
// See docs/design/plugin-providers.md.

// A SCOPED import: plugin exports a `Json` of its own, and the only name
// this file wants from it is `Definition`.
import struct VoxgigPlugin.Definition

/// Every plugin definition this library ships, in chain-documentation
/// order: the vaults, then the clouds, then the CLIs.
public let allplugins: [Definition] = [
  hashicorp, boru, awssecrets, awsparams, gcpsecrets, azuresecrets,
  onepassword, doppler, infisical, secretspec,
]
